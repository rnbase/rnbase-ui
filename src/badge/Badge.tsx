import React, { useEffect, useState } from 'react';
import {
  Animated,
  Easing,
  PixelRatio,
  StyleProp,
  StyleSheet,
  Text,
  TextStyle,
  ViewProps,
  ViewStyle,
} from 'react-native';

import { Themed, Theme, WithThemeProps, withTheme } from '../theming';

export interface BadgeProps extends ViewProps {
  size?: number;
  color?: string;
  value?: number | boolean;
  limit?: number | boolean;
  rounded?: boolean;
  animate?: boolean;
  style?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  themeKey?: string;
}

const Badge: React.FC<Themed<typeof createStyleSheet, BadgeProps>> = ({
  styles,
  size = 12,
  color,
  value,
  limit = 99,
  rounded = true,
  animate = true,
  style,
  textStyle,
  ...props
}) => {
  const toValue = value ? 1 : 0;
  const [animatedValue] = useState(() => new Animated.Value(toValue));

  useEffect(() => {
    Animated.timing(animatedValue, {
      duration: 300,
      toValue: toValue,
      useNativeDriver: true,
      easing: Easing.in(Easing.elastic(1)),
    }).start();
  }, [animatedValue, toValue]);

  if (!value) {
    return null;
  }

  const rootStyles = [
    styles.root,
    style,
    animate && {
      transform: [{ scale: animatedValue }],
    },
  ];

  if (color) {
    rootStyles.push({ backgroundColor: color });
  }

  let content;
  let height = styles.root.minHeight;

  if (typeof value === 'number' && value !== 0) {
    const displayValue = limit && value > limit ? `${limit}+` : value;

    height = PixelRatio.roundToNearestPixel(size / 0.6);

    rootStyles.push({
      height,
      minWidth: height,
      paddingHorizontal: size / 2,
    });

    const textStyles = [
      styles.text,
      textStyle,
      {
        fontSize: size,
      },
    ];

    content = (
      <Text style={textStyles} numberOfLines={1}>
        {displayValue}
      </Text>
    );
  }

  if (rounded) {
    rootStyles.push({ borderRadius: height / 2 });
  }

  return (
    <Animated.View {...props} style={rootStyles}>
      {content}
    </Animated.View>
  );
};

const createStyleSheet = ({ Colors, Fonts }: Theme) =>
  StyleSheet.create({
    root: {
      minWidth: 10,
      minHeight: 10,
      alignSelf: 'center',
      alignItems: 'center',
      justifyContent: 'center',
      backgroundColor: Colors.blue,
    },
    text: {
      ...Fonts.normal,
      color: Colors.white,
    },
  });

export type Props = BadgeProps & WithThemeProps;

export default withTheme(Badge, createStyleSheet, 'Badge');
