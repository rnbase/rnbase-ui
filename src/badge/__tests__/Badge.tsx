import React from 'react';
import { Animated as AnimatedObj } from 'react-native';

// Note: test renderer must be required after react-native.
import TestRenderer, { act } from 'react-test-renderer';

import { Props } from '../Badge';

jest.doMock('react-native/Libraries/Animated/src/Animated', () => {
  const Animated = jest.requireActual('react-native/Libraries/Animated/src/Animated');

  return {
    ...Animated,
    spring: jest.fn((value: any, { toValue }: any) => {
      value.setValue(toValue);

      return {
        start: jest.fn(),
      };
    }),
  };
});

/**
 * Under test
 */
const { default: Badge } = jest.requireActual('../Badge');

const Animated: any = AnimatedObj;

const createElement = (props: Props) => <Badge {...props} />;

const createRenderer = (props: Props) => TestRenderer.create(createElement(props));

it('should render normally', () => {
  expect(createRenderer({ value: 12 })).toMatchSnapshot();
});

it('should render without number', () => {
  expect(createRenderer({ value: true })).toMatchSnapshot();
});

it('should render when greather than limit', () => {
  expect(createRenderer({ value: 500, limit: 99 })).toMatchSnapshot();
});

it('should render with custom props', () => {
  expect(createRenderer({ value: 12, size: 16, cornerRadius: 0 })).toMatchSnapshot();
});

it('should not render when value = 0', () => {
  expect(createRenderer({ value: 0 })).toMatchSnapshot();
});

it('should not render when value = undefined', () => {
  expect(createRenderer({})).toMatchSnapshot();
});

describe('animation', () => {
  const start = jest.fn();

  Animated.spring.mockImplementation(() => ({ start }));

  beforeEach(() => {
    start.mockClear();
  });

  it('should not animate', () => {
    const tree = createRenderer({ value: 1 });

    act(() => {});

    Animated.spring.mockClear();

    act(() => {
      tree.update(createElement({ value: 2 }));
    });

    expect(Animated.spring).not.toBeCalled();
  });

  it('should animate appearance', () => {
    const tree = createRenderer({ value: 0 });

    act(() => {});

    Animated.spring.mockClear();

    act(() => {
      tree.update(createElement({ value: 1 }));
    });

    expect(Animated.spring).toBeCalledWith(expect.objectContaining({ _value: 0 }), {
      toValue: 1,
      friction: 6,
      tension: 60,
      useNativeDriver: true,
    });
    expect(start).toBeCalled();
  });
});
