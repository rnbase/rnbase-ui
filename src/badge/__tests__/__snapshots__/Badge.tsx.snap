// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not render when value = 0 1`] = `null`;

exports[`should not render when value = undefined 1`] = `null`;

exports[`should render normally 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": "#007aff",
      "borderRadius": 10,
      "height": 20,
      "justifyContent": "center",
      "minHeight": 10,
      "minWidth": 20,
      "paddingHorizontal": 6,
      "transform": Array [
        Object {
          "scale": 1,
        },
      ],
    }
  }
>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "color": "#ffffff",
          "fontFamily": "System",
          "fontWeight": "400",
          "includeFontPadding": false,
          "textAlignVertical": "center",
        },
        undefined,
        Object {
          "fontSize": 12,
        },
      ]
    }
  >
    12
  </Text>
</View>
`;

exports[`should render when greather than limit 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": "#007aff",
      "borderRadius": 10,
      "height": 20,
      "justifyContent": "center",
      "minHeight": 10,
      "minWidth": 20,
      "paddingHorizontal": 6,
      "transform": Array [
        Object {
          "scale": 1,
        },
      ],
    }
  }
>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "color": "#ffffff",
          "fontFamily": "System",
          "fontWeight": "400",
          "includeFontPadding": false,
          "textAlignVertical": "center",
        },
        undefined,
        Object {
          "fontSize": 12,
        },
      ]
    }
  >
    99+
  </Text>
</View>
`;

exports[`should render with custom props 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": "red",
      "height": 26.5,
      "justifyContent": "center",
      "minHeight": 10,
      "minWidth": 26.5,
      "paddingHorizontal": 8,
      "transform": Array [
        Object {
          "scale": 1,
        },
      ],
    }
  }
>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "color": "#ffffff",
          "fontFamily": "System",
          "fontWeight": "400",
          "includeFontPadding": false,
          "textAlignVertical": "center",
        },
        undefined,
        Object {
          "fontSize": 16,
        },
      ]
    }
  >
    12
  </Text>
</View>
`;

exports[`should render without number 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": "#007aff",
      "borderRadius": 5,
      "justifyContent": "center",
      "minHeight": 10,
      "minWidth": 10,
      "transform": Array [
        Object {
          "scale": 1,
        },
      ],
    }
  }
/>
`;
